File: src/Coq/Intf/ConstrExpr.hs
module Coq.Intf.ConstrExpr where

import Coq

import Coq.Intf.DeclKinds
import Coq.Kernel.Names
import Coq.Lib.Loc
import Coq.Library.GlobNames
import Coq.Library.LibNames

type Notation = String

data Explicitation
  = ExplByPos Int (Maybe Id)
  | ExplByName Id

data BinderKind
  = Default BindingKind
  | Generalized BindingKind BindingKind Bool
  deriving (Show)

data AbstractionKind
  = AbsLambda
  | AbsPi
  deriving (Show)

type ProjFlag = Maybe Int

data PrimToken
  = Numeral Integer
  | String String
  deriving (Show)

data RawCasesPatternExpr
  = RCPatAlias Loc RawCasesPatternExpr Id
  | RCPatCstr Loc GlobalReference [RawCasesPatternExpr] [RawCasesPatternExpr]
  | RCPatAtom Loc (Maybe Id)
  | RCPatOr Loc [RawCasesPatternExpr]
  deriving (Show)

data CasesPatternExpr
  = CPatAlias Loc CasesPatternExpr Id
  | CPatCstr Loc Reference [CasesPatternExpr] [CasesPatternExpr]
  | CPatAtom Loc (Maybe Reference)
  | CPatOr Loc [CasesPatternExpr]
  | CPatNotation Loc Notation CasesPatternNotationSubstitution [CasesPatternExpr]
  | CPatPrim Loc PrimToken
  | CPatRecord Loc [(Reference, CasesPatternExpr)]
  | CPatDelimiters Loc String CasesPatternExpr

type CasesPatternNotationSubstitution = ([CasesPatternExpr], [[CasesPatternExpr]])

type InstanceExpr = [GlobLevel]

data ConstrExpr
  = CRef Reference (Maybe InstanceExpr)
  | CFix Loc (Located Id) [FixExpr]
  | CCoFix Loc (Located Id) [CoFixExpr]
  | CProdN Loc [BinderExpr] ConstrExpr
  | CLambdaN Loc [BinderExpr] ConstrExpr
  | CLetIn Loc (Located Name) ConstrExpr ConstrExpr
  | CAppExpl Loc (ProjFlag, Reference, Maybe InstanceExpr) [ConstrExpr]
  | CApp Loc (ProjFlag, ConstrExpr) [(ConstrExpr, Maybe Explicitation)]
  | CRecord Loc (Maybe ConstrExpr) [(Reference, ConstrExpr)]
  | CCases Loc CaseStyle (Maybe ConstrExpr) [CaseExpr] [BranchExpr]
  | CLetTuple Loc [Located Name] (Maybe (Located Name), Maybe ConstrExpr) ConstrExpr ConstrExpr
  | CIf Loc ConstrExpr (Maybe (Located Name), Maybe ConstrExpr) ConstrExpr ConstrExpr
  | CHole Loc (Maybe EvarKinds) IntroPatternNamingExpr (Maybe RawGenericArgument)
  | CPatVar Loc PatVar
  | CEvar Loc ExistentialName [(Id, ConstrExpr)]
  | CSort Loc GlobSort
  | CCast Loc ConstrExpr (CastType ConstrExpr)
  | CNotation Loc Notation ConstrNotationSubstitution
  | CGeneralization Loc BindingKind (Maybe AbstractionKind) ConstrExpr
  | CPrim Loc PrimToken
  | CDelimiters Loc String ConstrExpr
  deriving (Show)

type CaseExpr = (ConstrExpr, (Maybe (Located Name), Maybe CasesPatternExpr))

type BranchExpr = (Loc, [Located [CasesPatternExpr]], ConstrExpr)

type BinderExpr = ([Located Name], BinderKind, ConstrExpr)

type FixExpr = (Located Id, (Maybe (Located Id), RecursionOrderExpr), [LocalBinder], ConstrExpr, ConstrExpr)

type CoFixExpr = (Located Id, [LocalBinder], ConstrExpr, ConstrExpr)

data RecursionOrderExpr
  = CStructRec
  | CWfRec ConstrExpr
  | CMeasureRec ConstrExpr (Maybe ConstrExpr)
  deriving (Show)

data LocalBinder
  = LocalRawDef (Located Name) ConstrExpr
  | LocalRawAssum [Located Name] BinderKind ConstrExpr
  deriving (Show)

type ConstrNotationSubstitution = ([ConstrExpr], [[ConstrExpr]], [[LocalBinder]])

type TypeClassConstraint = (Located Name, BindingKind, ConstrExpr)

type TypeClassContext = [TypeClassConstraint]

type ConstrPatternExpr = ConstrExpr

data WithDeclarationAST
  = CWith_Module (Located [Id]) (Located QualId)
  | CWith_Definition (Located [Id]) ConstrExpr
  deriving (Show)

data ModuleAST
  = CMident (Located QualId)
  | CMapply Loc ModuleAST ModuleAST
  | CMwith Loc ModuleAST WithDeclarationAST
  deriving (Show)

File: web/ts/coq/intf/constr-expr.ts
import * as _ from 'lodash'

import * as CaseStyle from '../case-style'
import * as EvarKinds from './evar-kinds'
import * as MiscTypes from './misctypes'
import * as PpExtend from '../interp/ppextend'
import { C_AST } from '../lib/c-ast'
import * as GenArg from '../lib/genarg'
import { Located } from '../lib/loc'
import * as LibNames from '../library/libnames'

export type ConstrExprR
    = CRef
    | CFix
    | CCoFix
    | CProdN
    | CLambdaN
    | CLetIn
    | CAppExpl
    | CApp
// | CRecord
    | CCases
    | CLetTuple
// | CIf
    | CHole
// | CPatVar
// | CEvar
    | CSort
// | CCast
    | CNotation
// | CGeneralization
    | CPrim
    | CDelimiters

export type ConstrExpr = C_AST<ConstrExprR>

export type ConstrNotationSubstitution = [
    ConstrExpr[],
    ConstrExpr[][],
    CasesPatternExpr[],
    LocalBinderExpr[][]
]

type ProjFlag = Maybe<number>

type AppFun = [ProjFlag, ConstrExpr]
export type AppArg = [ConstrExpr, Maybe<Located<Explicitation>>]
export type AppArgs = AppArg[]

export class CApp {
    constructor(
        public funct : AppFun,
        public args : AppArgs
    ) { }
}

type AppExplFun = [ProjFlag, LibNames.Reference, Maybe<InstanceExpr>]
export class CAppExpl {
    constructor(
        public readonly funct : AppExplFun,
        public readonly args : ConstrExpr[]
    ) { }
}

type CaseExpr = [
    ConstrExpr,
    Maybe<MiscTypes.lname>,
    Maybe<CasesPatternExpr>
]

export type BranchExpr = C_AST<[
    CasesPatternExpr[][],
    ConstrExpr
]>

export class CCases {
    constructor(
        public caseStyle : CaseStyle.CaseStyle,
        public returnType : Maybe<ConstrExpr>,
        public cases : CaseExpr[],
        public branches : BranchExpr[]
    ) { }
}

export class CCoFix {
    private tag : 'CCoFix'
    // TODO
}

export class CDelimiters {
    constructor(
        public str : string,
        public expr : ConstrExpr
    ) { }
}

export class CFix {
    private tag : 'CFix'
    constructor(
    ) { }
}

export class CHole {
    constructor(
        public evarKinds : Maybe<EvarKinds.T>,
        public introPatternNamingExpr : MiscTypes.IntroPatternNamingExpr,
        public rawGenericArgument : Maybe<GenArg.RawGenericArgument>,
    ) { }
}

export class CLambdaN {
    constructor(
        public binders : LocalBinderExpr[],
        public body : ConstrExpr
    ) { }
}

export class CLetIn {
    constructor(
        public name : MiscTypes.lname,
        public bound : ConstrExpr,
        public type : Maybe<ConstrExpr>,
        public body : ConstrExpr,
    ) { }
}

export class CLetTuple {
    constructor(
        public names : MiscTypes.lname[],
        public returnType : [Maybe<MiscTypes.lname>, Maybe<ConstrExpr>],
        public bound : ConstrExpr,
        public body : ConstrExpr
    ) { }
}

export type Notation = String

export class CNotation {
    constructor(
        public notation : Notation,
        public substitution : ConstrNotationSubstitution,
        public precedence : number,
        public unparsing : PpExtend.Unparsing[]
    ) { }
}

export class CProdN {
    constructor(
        public binderList : LocalBinderExpr[],
        public returnExpr : ConstrExpr
    ) { }
}

export class CPrim {
    constructor(
        public token : PrimToken
    ) { }
}

export class CRef {
    constructor(
        public reference : LibNames.Reference,
        public universeInstance : Maybe<InstanceExpr>
    ) { }
}

export class CSort {
    constructor(
        public globSort : GlobSort
    ) { }
}

export type LocalBinderExpr
    = CLocalAssum
    | CLocalDef
    | CLocalPattern

export class CLocalAssum {
    constructor(
        public readonly names : MiscTypes.lname[],
        public readonly binderKind : BinderKind,
        public readonly type : ConstrExpr,
    ) { }
}

export class CLocalDef {
    constructor(
        public readonly name : MiscTypes.lname,
        public readonly value : ConstrExpr,
        public readonly optionalType : Maybe<ConstrExpr>,
    ) { }
}

export class CLocalPattern {
    constructor(
        public readonly pattern : C_AST<[CasesPatternExpr, Maybe<ConstrExpr>]>,
    ) { }
}

// added for PeaCoq
// export function extractProdBinders(a : ConstrExprR) : [LocalBinderExpr[], ConstrExprR] {
//     if (a instanceof CProdN) {
//         const [bl, c] : [any[], any] = [a.binderList, a.returnExpr]
//         if (bl.length === 0) {
//             return extractProdBinders(a.returnExpr)
//         } else {
//             const [nal, bk, t] = bl[0]
//             const [blrec, cRest] = extractProdBinders(new CProdN(_.tail(bl), c))
//             const l : LocalBinderExpr[] = [new CLocalAssum(nal, bk, t)]
//             return [l.concat(blrec), cRest]
//         }
//     }
//     return [[], a]
// }

// added for PeaCoq, much less useful than in the past now
// export function extractLamBinders(a : ConstrExprR) : [LocalBinderExpr[], ConstrExprR] {
//     if (a instanceof CLambdaN) {
//         const [lamBinders, rest] = extractLamBinders(a.body)
//         return [
//             ([] as LocalBinderExpr[]).concat(a.binders, lamBinders),
//             rest
//         ]
//     } else {
//         return [[], a]
//     }
// }

export type CasesPatternExpr = C_AST<CasesPatternExprR>

export type CasesPatternExprR
    = CPatCstr
    | CPatAtom
    | CPatNotation
    | CPatPrim
    | CPatDelimiters

// TODO CPatAlias

export class CPatCstr {
    constructor(
        public reference : LibNames.Reference,
        public cases1 : Maybe<CasesPatternExpr[]>,
        public cases2 : CasesPatternExpr[]
    ) { }
}

export class CPatAtom {
    constructor(
        public reference : Maybe<LibNames.Reference>
    ) { }
}

// TODO CPatAtom
// TODO CPatOr

type CasesPatternNotationSubstitution = [
    CasesPatternExpr[],
    CasesPatternExpr[][]
]

export class CPatNotation {
    constructor(
        public notation : Notation,
        public substitution : CasesPatternNotationSubstitution,
        public patterns : CasesPatternExpr[],
        // the next fields are added by PeaCoq to simplify things
        public precedence : number,
        public unparsing : PpExtend.Unparsing[]
    ) { }
}
// TODO CPatNotation

export class CPatPrim {
    constructor(
        public token : PrimToken
    ) { }
}

// TODO CPatRecord

export class CPatDelimiters {
    constructor(
        public str : string,
        public cases : CasesPatternExpr
    ) { }
}
